### ISSUE TITLE
DP utitilies on Int and Float

### ISSUE THREAD

**Issue by imminent** [android/android-ktx#132]
Calculating dimensions, such as, SP and DP require the `DisplayMetrics`, so you can't do something too simple without - say - adding Extensions inside of `Activity` and such directly. But you can still make some progress similar to the `Duration` extensions doing something like this:
```kotlin
/**
 * Applies density dimensions to the [Int]
 */
fun Int.dp(displayMetrics: DisplayMetrics) = toFloat().dp(displayMetrics).toInt()

/**
 * Applies density dimensions to the [Float]
 */
fun Float.dp(displayMetrics: DisplayMetrics) = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, this, displayMetrics)
```

Which would be called like `9.dp(displayMetrics)`.

**Comment by JakeWharton** [comment]
I think we would call these `toDp` as conversion factory extensions are prefixed with "to".

**Comment by imminent** [comment]
Ah, that's fine, updating the description to `to`. 👍

**Comment by nohitme** [comment]
Just curious, why is it applied on Int & Float instead of on DisplayMetrics ike `displayMetrics.toDp(9)`?

That way you can also name the param "pixel" so it's less confusing.

**Comment by daberni** [comment]
I am not sure if it should be treated/interpreted as a conversion. mostly you simply want do pass a value in dp to a method which takes pixels. so `9.toDp()` I would assume converting `9px` to `Ndp` but what we want is to be get the value `9` get treated as `dp` and therefore I would prefer `9.asDp()`

**Comment by jacklt** [comment]
I wrote this utility:

```
fun Context.toPixelFromDip(value: Float) = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, value, resources.displayMetrics)
```

```
fun View.toPixelFromDip(value: Float) = context.toPixelFromDip(value)
```

Pro:
- `toPixelFromDip()` is very clear the origin and the destination
- It's handy when you call this method from a `Context` or (more common) a `View`, ex: `View().apply { translationX = toPixelFromDip(2) }`

**Comment by lupajz** [comment]
Wouldn't accessing system `Resources` just be sufficiet? So we can have something like:
```
val Int.dp: Int // or just return Float here
    get() = (this * Resources.getSystem().displayMetrics.density).toInt()
```
and use it in place like `val side = 42.dp` ?

**Comment by jacklt** [comment]
I don't think so. Your application resources may have a different density. Maybe not now, but you don't know in future.
Also seems correct take it from application context.

**Comment by romainguy** [comment]
@johanneslaubermoovel This doesn't work with negative numbers. Also you are relying on display metrics that might not match the display.

**Comment by hendraanggrian** [comment]
I am against using `Number` as input since only `Int` and `Float` are the common types representing dimensions in Android.

> This wouldn't work with multiple displays (virtual or physical).

@romainguy this just blew my mind. Say a secondary display is connected, which `DisplayMetrics` then would the `Resources` refers to?

**Comment by romainguy** [comment]
`Resources` will use the `DisplayMetrics` of the `Display` the `Resources` came from.

**Comment by lukewaggoner** [comment]
I personally would go with the following methods:

    fun Number.toDp(displayMetrics: DisplayMetrics? = null): Float = this.toFloat() / ((displayMetrics ?: Resources.getSystem().displayMetrics).densityDpi / 160f)
    fun Number.toPx(displayMetrics: DisplayMetrics? = null) : Float = this.toFloat() / ((displayMetrics ?: Resources.getSystem().displayMetrics).densityDpi / 160f)

`displayMetrics` is an optional parameter in case you have multiple displays. If not, you have the option of a simple `8.px()` or `12.px()`.

**Comment by romainguy** [comment]
This seems error prone. The DisplayMetrics should be mandatory. BTW no need to / 160.0f, just use `DisplayMetrics.density` instead (and you need to properly round and handle negative values correctly).

**Comment by iglaweb** [comment]
@JakeWharton What about using these converters?

```kotlin
fun Number.toDp(displayMetrics: DisplayMetrics): Float = this.toFloat() / displayMetrics.density
fun Number.toSp(displayMetrics: DisplayMetrics): Float = this.toFloat() / displayMetrics.scaledDensity
fun Number.toPx(displayMetrics: DisplayMetrics, fromSp: Boolean = false): Float
        = this.toFloat() * (if (fromSp) displayMetrics.scaledDensity else displayMetrics.density)
```

**Comment by romainguy** [comment]
These are incorrect, `toPx` in particular does not round properly (which needs to take negative numbers into account). I would also prefer to not have a `fromSp` boolean, but a separate method.

**Comment by imminent** [comment]
If you allow arch components, then you can have an onCreateProperty (which is like an Android-specific `lateinit var`) and define extensions on `Activity` and `Fragment`:

```kotlin
/**
 * Initializes a property with the DP value
 * @receiver An [Activity] that is a [LifecycleOwner]
 * @return [Float] representation of the DP value
 * @see [onCreate]
 */
fun <T> T.dp(value: Float) where T : Activity, T : LifecycleOwner = onCreate {
    value.toDp(resources.displayMetrics)
}

/**
 * Initializes a property with the DP value
 * @receiver A [Fragment] that is a [LifecycleOwner]
 * @return [Int] representation of the DP value
 * @see [onCreate]
 */
fun <T> T.dp(value: Int) where T : Fragment, T : LifecycleOwner = onCreate {
    value.toDp(resources.displayMetrics)
}

// In Activity
val margins by dp(5)
```

**Comment by iglaweb** [comment]
@romainguy I refined concrete types for these density functions, could you explain why they are not applicable for negative values? (tested myself)

```kotlin
fun Int.toDp(displayMetrics: DisplayMetrics): Float = this.toFloat() / displayMetrics.density
fun Int.toSp(displayMetrics: DisplayMetrics): Float = this.toFloat() / displayMetrics.scaledDensity
fun Float.spToPx(displayMetrics: DisplayMetrics): Int =
        (this * displayMetrics.scaledDensity).roundToInt()
fun Float.dpToPx(displayMetrics: DisplayMetrics): Int =
        (this * displayMetrics.density).roundToInt()
```

**Comment by romainguy** [comment]
`roundToInt()` rounds toward the nearest integer (towards positive infinity when there's a tie), so it won't round in the right direction for negative numbers. For instance `0.5f` rounds to 1, but `-0.5f` rounds to 0.

**Comment by iglaweb** [comment]
@romainguy now round should be ok

```kotlin
fun Float.spToPx(displayMetrics: DisplayMetrics): Int =
        (this * displayMetrics.scaledDensity).round()
fun Float.dpToPx(displayMetrics: DisplayMetrics): Int =
        (this * displayMetrics.density).round()
private fun Float.round(): Int = (if(this < 0) ceil(this - 0.5f) else floor(this + 0.5f)).toInt()
```

**Comment by arekolek** [comment]
How about a small DSL that would make it possible to write stuff like:

```kotlin
view.context.pixelConversions {
    outRect.left = 80.dp
    outRect.right = 20.dp
}
```

Using the idea from the first post, that would be:

```kotlin
fun Context.pixelConversions(block: ConversionContext.() -> Unit) {
    block(ConversionContext(resources.displayMetrics))
}

class ConversionContext(private val displayMetrics: DisplayMetrics) {
    val Int.dp get() = toFloat().dp.toInt()

    val Float.dp get() = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, this, displayMetrics)
}
```

After trying to apply it in a different setting, I'm not sure how useful this idea actually is 🤔 

(I forgot this repo is unused now, is this ticket tracked on google issue tracker?)

**Comment by wlara** [comment]
For convenience, can all of these functions use the System display metrics by default? Example:
```
fun Int.toDp(displayMetrics: DisplayMetrics = Resources.getSystem().displayMetrics)
```
